src/controllers/auth.ts

```diff
@@ -0,0 +1,235 @@
+        if (existingUserByEmail) {
+            const CONFLICT_STATUS = 409;
+            res.status(CONFLICT_STATUS).send("Email already in use");
+            return;
+        }
+
+        const existingUserByUsername = await userModel.findOne({ username: req.body.username });
+        if (existingUserByUsername) {
+            const CONFLICT_STATUS = 409;
+            res.status(CONFLICT_STATUS).send("Username already in use");
+            return;
+        }
+
+        const { password } = req.body;
+        const SALT_ROUNDS = 10;
+        const salt = await bcrypt.genSalt(SALT_ROUNDS);
+
+        const hashedPassword = await bcrypt.hash(password, salt);
+        const user = await userModel.create({
+            email: req.body.email,
+            password: hashedPassword,
+            username: req.body.username,
+        });
+        res.status(200).send({
+            _id: user._id,
+            email: user.email,
+            username: user.username,
+        });
+    } catch (err: any) {
+        const INTERNAL_SERVER_ERROR = 500;
+        res.status(INTERNAL_SERVER_ERROR).send("Server Error");
+    }
+};
+
+interface tTokens {
+    accessToken: string
+    refreshToken: string
+}
+
+const generateToken = (userId: string): tTokens | null => {
+    if (!tokenSecret) {
+        return null;
+    }
+    // Generate token
+    const random = Math.random().toString();
+    const accessToken = jwt.sign({
+        _id: userId,
+        random,
+    },
+    tokenSecret,
+    { expiresIn: tokenExpires });
+
+    const refreshToken = jwt.sign({
+        _id: userId,
+        random,
+    },
+    tokenSecret,
+    { expiresIn: refreshTokenExpires });
+    return {
+        accessToken,
+        refreshToken,
+    };
+};
+// eslint-disable-next-line max-statements
+const login = async (req: Request, res: Response) => {
+    try {
+        const user = await userModel.findOne({ email: req.body.email });
+        if (!user) {
+            res.status(400).send("wrong username or password");
+            return;
+        }
+        const validPassword = await bcrypt.compare(req.body.password, user.password);
+        if (!validPassword) {
+            res.status(400).send("wrong username or password");
+            return;
+        }
+        if (!tokenSecret) {
+            res.status(500).send("Server Error");
+            return;
+        }
+        // Generate token
+        const tokens = generateToken(user._id);
+        if (!tokens) {
+            res.status(500).send("Server Error");
+            return;
+        }
+        if (!user.refreshToken) {
+            user.refreshToken = [];
+        }
+        user.refreshToken.push(tokens.refreshToken);
+        await user.save();
+        res.status(200).send(
+            {
+                _id: user._id,
+                accessToken: tokens.accessToken,
+                refreshToken: tokens.refreshToken,
+            });
+    } catch (err) {
```

Comment:
This check is unnecessary. If an unexpected error occurred, the default error handler will return 500.